app/controllers/orders_controller.rb

```diff
@@ -0,0 +1,186 @@
+  before_action :find_order, only: [:confirmation]
+
+  def index
+  end
+
+  def create
+  end
+
+  def checkout_form
+    @title = "Checkout"
+  end
+
+  def checkout
+    #verify everything is in stock
+    @cart.orderitems.each do |orderitem|
+      if orderitem.product.quantity == 0
+        flash[:status] = :error
+        flash[:result_text] = "#{orderitem.product.name} is out of stock"
+        return redirect_to show_cart_path
+      elsif orderitem.quantity > orderitem.product.quantity
+        flash[:status] = :error
+        flash[:result_text] = "You attempted to purchase #{orderitem.quantity} #{orderitem.product.name}, but there are only #{orderitem.product.quantity} available."
+        return redirect_to show_cart_path
+      end
+    end
+    #validate user input
+
+    #modify the cart
+    @cart.status = "paid"
+    @cart.purchase_datetime = DateTime.now
+    puts @cart.purchase_datetime.to_date
+
+    @cart.update_attributes(checkout_params)
+
+    #checkout the cart
+    if @cart.save
+      @cart.orderitems.each do |orderitem|
+        product = orderitem.product
+        product.quantity -= orderitem.quantity
+        product.save
+      end
+
+      session[:order_id] = nil
+      flash[:status] = :success
+      flash[:result_text] = "Your order has been placed"
+      return redirect_to order_confirmation_path(@cart.id)
+    else
+      flash.now[:status] = :error
+      flash.now[:result_text] = "Your order could not be placed at this time"
+      flash.now[:messages] = @cart.errors.messages
+      render :checkout_form
+    end
+
+  end #checkout
+
+  def confirmation
+    @total = @order.orderitems.sum { |orderitem| (orderitem.quantity * orderitem.product.price) }
+    @title = "Thank You"
+  end
+
+  def new
+  end
+
+  def edit
+  end
+
+  def update
+  end
+
+  def show
+  end
+
+  def show_cart
+    @title = "My Cart"
+  end
+
+  def add_item
+    product = Product.find_by(id: params[:product_id])
+    if product
+      if product.quantity >= 1
+        if @order_item
+          @order_item.quantity += 1
+        else
+          @order_item = Orderitem.new(order_id: session[:order_id], product_id: params[:product_id], quantity: 1)
+        end
+        if @order_item.save
+          flash[:status] = :success
+          flash[:result_text] = "1 #{product.name} added to your cart"
+          redirect_back(fallback_location: root_path)
+        else
+          flash[:status] = :error
+          flash[:result_text] = "Unable to add #{product.name} to your cart"
+          redirect_back(fallback_location: root_path)
+        end
+      else
+        flash[:status] = :error
+        flash[:result_text] = "#{product.name} is out of stock"
+        redirect_back(fallback_location: root_path)
+      end
+    else
+      flash[:status] = :error
+      flash[:result_text] = "That is not a valid product"
+      redirect_back(fallback_location: root_path)
+    end
+  end
+
+  def update_quantity
+    if @order_item
```

Comment:
in this controller, you check for the truthiness of `@order_item` a couple of times, and if `@order_item` is not truthy, it shows a specific (but repeated) error message and redirection. You can handle this in the controller filter `find_orderitem` that you made and reduce duplicate logic.

Same goes for a couple of the other instance variables in this controller where it makes sense